test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,769 @@
+*/
+
+
+public class SinceChecker {
+    private final Map<String, Set<String>> LEGACY_PREVIEW_METHODS = new HashMap<>();
+    private final Map<String, IntroducedIn> classDictionary = new HashMap<>();
+    private final JavaCompiler tool;
+    private final List<String> wrongTagsList = new ArrayList<>();
+
+    public static void main(String[] args) throws Exception {
+        if (args.length == 0) {
+            throw new SkippedException("Test module not specified");
+        }
+        SinceChecker sinceCheckerTestHelper = new SinceChecker();
+        sinceCheckerTestHelper.testThisModule(args[0]);
+    }
+
+    private SinceChecker() throws IOException {
+        tool = ToolProvider.getSystemJavaCompiler();
+        for (int i = 9; i <= Runtime.version().feature(); i++) {
+            JavacTask ct = (JavacTask) tool.getTask(null, null, null,
+                    List.of("--release", String.valueOf(i)),
+                    null,
+                    Collections.singletonList(SimpleJavaFileObject.forSource(URI.create("myfo:/Test.java"), "")));
+            ct.analyze();
+
+            String version = String.valueOf(i);
+            Elements elements = ct.getElements();
+            elements.getModuleElement("java.base"); // forces module graph to be instantiated
+            elements.getAllModuleElements().forEach(me ->
+                    processModuleRecord(me, version, ct));
+        }
+    }
+
+    private void processModuleRecord(ModuleElement moduleElement, String releaseVersion, JavacTask ct) {
+        for (ModuleElement.ExportsDirective ed : ElementFilter.exportsIn(moduleElement.getDirectives())) {
+            persistElement(moduleElement, moduleElement, ct.getTypes(), releaseVersion);
+            if (ed.getTargetModules() == null) {
+                analyzePackageRecord(ed.getPackage(), releaseVersion, ct);
+            }
+        }
+    }
+
+    private void analyzePackageRecord(PackageElement pe, String releaseVersion, JavacTask ct) {
+        persistElement(pe, pe, ct.getTypes(), releaseVersion);
+        List<TypeElement> typeElements = ElementFilter.typesIn(pe.getEnclosedElements());
+        for (TypeElement te : typeElements) {
+            analyzeClassRecord(te, releaseVersion, ct.getTypes(), ct.getElements());
+        }
+    }
+
+    private void analyzeClassRecord(TypeElement te, String version, Types types, Elements elements) {
+        Set<Modifier> classModifiers = te.getModifiers();
+        if (!(classModifiers.contains(Modifier.PUBLIC) || classModifiers.contains(Modifier.PROTECTED))) {
+            return;
+        }
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(element -> element.getModifiers().contains(Modifier.PUBLIC) || element.getModifiers().contains(Modifier.PROTECTED))
```

Comment:
Consider writing and using a predicate method:

```
boolean isDocumented(Element e) {
    var mods = e.getModifiers();
    return mods.contains(Modifier,.PUBLIC) || mods.contains(Modifier.PROTECTED);
}
```

You could then use that method in the obvious way for the class declaration(line 134) and in a lambda as follows:
```
    .filter(this::isDocumented)
```

Reply:
Added in 32edb4d811978f87bd2f850c214191e8fc7b8c88